File: src/python/bc_calculator_mcp/bc_process.py
"""
BC Process wrapper - manages individual BC calculator process
Handles process lifecycle, I/O communication, and error handling
"""

import asyncio
import sys
from asyncio.subprocess import Process
from typing import Optional

from .types import BCCalculatorError, BCErrorCode, BCProcessOptions


class BCProcess:
    """Wrapper for a single BC calculator process"""
    
    def __init__(self, options: BCProcessOptions) -> None:
        self._options = options
        self._process: Optional[Process] = None
        self._is_ready: bool = False
        self._current_precision: int = options.precision
        self._lock: asyncio.Lock = asyncio.Lock()
        self._pending_request: bool = False
        self._output_buffer: str = ""
        self._error_buffer: str = ""
    
    async def start(self) -> None:
        """Start the BC process and initialize it"""
        try:
            # Spawn BC with math library (-l flag)
            # Important: No shell for security
            self._process = await asyncio.create_subprocess_exec(
                'bc', '-l',
                stdin=asyncio.subprocess.PIPE,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE
            )
            
            # Wait a moment for process to be ready
            await asyncio.sleep(0.1)
            
            # Mark as ready
            self._is_ready = True
            
            # Initialize precision using BC's scale variable
            await self.set_precision(self._options.precision)
            
        except FileNotFoundError:
            raise BCCalculatorError(
                "BC calculator not found. Please install bc (e.g., 'sudo apt-get install bc')",
                BCErrorCode.BC_SPAWN_ERROR
            )
        except Exception as error:
            raise BCCalculatorError(
                f"Failed to start BC process: {str(error)}",
                BCErrorCode.BC_SPAWN_ERROR,
                error
            )
    
    async def evaluate(self, expression: str, timeout: Optional[int] = None) -> str:
        """
        Evaluate a mathematical expression using BC
        
        Args:
            expression: The expression to evaluate (already validated)
            timeout: Optional timeout in milliseconds (overrides default)
            
 Returns:
            The calculated result as a string
            
        Raises:
            BCCalculatorError: If process not ready, busy, or evaluation fails
        """
        if not self._is_ready or not self._process or not self._process.stdin:
            raise BCCalculatorError(
                "BC process not ready",
                BCErrorCode.BC_NOT_READY
            )
        
        async with self._lock:
            if self._pending_request:
                raise BCCalculatorError(
                    "BC process is busy with another calculation",
                    BCErrorCode.BC_NOT_READY
                )
            
            self._pending_request = True
            timeout_sec = (timeout or self._options.timeout) / 1000.0
            
            try:
                # Clear buffers
                self._output_buffer = ""
                self._error_buffer = ""
                
                # Send expression to BC stdin
                self._process.stdin.write((expression + '\n').encode())
                await self._process.stdin.drain()
                
                # Read result with timeout
                result = await asyncio.wait_for(
                    self._read_result(),
                    timeout=timeout_sec
                )
                
                return result
                
            except asyncio.TimeoutError:
                # Kill the hung process
                self.kill()
                raise BCCalculatorError(
                    f"Calculation timeout after {timeout_sec * 1000:.0f}ms",
                    BCErrorCode.BC_TIMEOUT
                )
            except Exception as error:
                raise BCCalculatorError(
                    f"Failed to evaluate expression: {str(error)}",
                    BCErrorCode.INTERNAL_ERROR,
                    error
                )
            finally:
                self._pending_request = False
    
    async def _read_result(self) -> str:
        """Read calculation result from stdout"""
        if not self._process or not self._process.stdout:
            raise BCCalculatorError(
                "BC process stdout not available",
                BCErrorCode.BC_NOT_READY
            )
        
        # Read until we get a complete line
        while True:
            # Check for errors first
            if self._process.stderr:
                try:
                    error_data = await asyncio.wait_for(
                        self._process.stderr.read(1024),
                        timeout=0.01
                    )
                    if error_data:
                        error_text = error_data.decode().strip()
                        if error_text:
                            raise BCCalculatorError(
                                f"BC error: {error_text}",
                                BCErrorCode.BC_RUNTIME_ERROR
                            )
                except asyncio.TimeoutError:
                    pass  # No error data available
            
            # Read stdout
            line = await self._process.stdout.readline()
            if not line:
                # Process ended unexpectedly
                raise BCCalculatorError(
                    "BC process ended unexpectedly",
                    BCErrorCode.BC_PROCESS_EXIT
                )
            
            decoded = line.decode().strip()
            if decoded:
                return decoded
    
    async def set_precision(self, scale: int) -> None:
        """
        Set the precision (decimal places) for calculations
        
        Args:
            scale: Number of decimal places (0-100)
            
        Raises:
            BCCalculatorError: If scale invalid or process not ready
        """
        # Validate scale
        if scale < 0 or scale > 100:
            raise BCCalculatorError(
                f"Invalid precision: {scale}. Must be between 0 and 100",
                BCErrorCode.VALIDATION_ERROR
            )
        
        if not self._process or not self._process.stdin:
            raise BCCalculatorError(
                "BC process not ready",
                BCErrorCode.BC_NOT_READY
            )
        
        # Set scale directly - BC's scale assignment produces no output
        self._process.stdin.write(f"scale={scale}\n".encode())
        await self._process.stdin.drain()
        self._current_precision = scale
        
        # Give BC a moment to process
        await asyncio.sleep(0.05)
    
    def get_precision(self) -> int:
        """Get current precision setting"""
        return self._current_precision
    
    def is_available(self) -> bool:
        """Check if the process is available for new calculations"""
        return (
            self._is_ready 
            and not self._pending_request 
            and self._process is not None
            and self._process.returncode is None
        )
    
    def kill(self) -> None:
        """Kill the BC process"""
        if self._process:
            try:
                self._process.kill()
            except Exception:
                pass  # Ignore errors when killing
            self._process = None
            self._is_ready = False
    
    def get_pid(self) -> Optional[int]:
        """Get process ID (for debugging)"""
        return self._process.pid if self._process else None

File: src/bc-process.ts
/**
 * BC Process wrapper - manages individual BC calculator process
 * Handles process lifecycle, I/O communication, and error handling
 */

import { spawn, ChildProcess } from 'child_process';
import { BCProcessOptions, BCCalculatorError, BCErrorCode } from './types.js';

interface PendingRequest {
  resolve: (value: string) => void;
  reject: (error: Error) => void;
  timer?: NodeJS.Timeout;
}

export class BCProcess {
  private process: ChildProcess | null = null;
  private isReady: boolean = false;
  private currentPrecision: number;
  private pendingRequest: PendingRequest | null = null;
  private outputBuffer: string = '';
  private errorBuffer: string = '';

  constructor(private options: BCProcessOptions) {
    this.currentPrecision = options.precision;
  }

  /**
   * Start the BC process and initialize it
   */
  async start(): Promise<void> {
    try {
      // Spawn BC with math library (-l flag)
      // Important: shell: false for security - no shell interpretation
      this.process = spawn('bc', ['-l'], {
        shell: false,
        stdio: ['pipe', 'pipe', 'pipe']
      });

      // Set up event handlers
      this.setupHandlers();

      // Wait a moment for process to be ready
      await new Promise(resolve => setTimeout(resolve, 100));

      // Mark as ready before setting precision
      this.isReady = true;

      // Initialize precision using BC's scale variable
      await this.setPrecision(this.options.precision);
    } catch (error) {
      throw new BCCalculatorError(
        `Failed to start BC process: ${error instanceof Error ? error.message : String(error)}`,
        BCErrorCode.BC_SPAWN_ERROR,
        error
      );
    }
  }

  /**
   * Set up event handlers for the BC process
   */
  private setupHandlers(): void {
    if (!this.process) return;

    // Handle stdout - collect calculation results
    this.process.stdout?.on('data', (data: Buffer) => {
      this.outputBuffer += data.toString();
      
      // BC outputs results line by line, ending with newline
      const lines = this.outputBuffer.split('\n');
      
      // If we have at least 2 lines (result + empty), we have a complete result
      if (lines.length >= 2) {
        // Keep the last incomplete line in buffer
        this.outputBuffer = lines.pop() || '';
        
        // Process complete lines (filter out empty lines)
        const results = lines.filter(line => line.trim());
        
        if (results.length > 0 && this.pendingRequest) {
          // Take the last result (most recent)
          const result = results[results.length - 1].trim();
          this.resolvePendingRequest(result);
        }
      }
    });

    // Handle stderr - capture errors from BC
    this.process.stderr?.on('data', (data: Buffer) => {
      const errorText = data.toString().trim();
      this.errorBuffer += errorText + '\n';
      
      // If we have a pending request and get an error, reject it
      if (errorText && this.pendingRequest) {
        this.rejectPendingRequest(
          new BCCalculatorError(
            `BC error: ${errorText}`,
            BCErrorCode.BC_RUNTIME_ERROR
          )
        );
      }
    });

    // Handle process exit
    this.process.on('exit', (code, signal) => {
      this.isReady = false;
      
      if (this.pendingRequest) {
        this.rejectPendingRequest(
          new BCCalculatorError(
            `BC process exited unexpectedly (code: ${code}, signal: ${signal})`,
            BCErrorCode.BC_PROCESS_EXIT,
            { code, signal }
          )
        );
      }
    });

    // Handle process errors
    this.process.on('error', (error) => {
      this.isReady = false;
      
      if (this.pendingRequest) {
        this.rejectPendingRequest(
          new BCCalculatorError(
            `BC process error: ${error.message}`,
            BCErrorCode.BC_SPAWN_ERROR,
            error
          )
        );
      }
    });
  }

  /**
   * Evaluate a mathematical expression using BC
   * @param expression The expression to evaluate (already validated)
   * @param timeout Optional timeout in milliseconds (overrides default)
   * @returns The calculated result as a string
   */
  async evaluate(expression: string, timeout?: number): Promise<string> {
    if (!this.isReady || !this.process || !this.process.stdin) {
      throw new BCCalculatorError(
        'BC process not ready',
        BCErrorCode.BC_NOT_READY
      );
    }

    if (this.pendingRequest) {
      throw new BCCalculatorError(
        'BC process is busy with another calculation',
        BCErrorCode.BC_NOT_READY
      );
    }

    const timeoutMs = timeout ?? this.options.timeout;

    return new Promise((resolve, reject) => {
      this.pendingRequest = { resolve, reject };
      
      // Clear buffers before new calculation
      this.outputBuffer = '';
      this.errorBuffer = '';

      // Set timeout
      this.pendingRequest.timer = setTimeout(() => {
        this.rejectPendingRequest(
          new BCCalculatorError(
            `Calculation timeout after ${timeoutMs}ms`,
            BCErrorCode.BC_TIMEOUT
          )
        );
        // Kill the hung process
        this.kill();
      }, timeoutMs);

      try {
        // Send expression to BC stdin
        // BC expects expressions to end with newline
        this.process!.stdin!.write(expression + '\n');
      } catch (error) {
        this.rejectPendingRequest(
          new BCCalculatorError(
            `Failed to write to BC process: ${error instanceof Error ? error.message : String(error)}`,
            BCErrorCode.INTERNAL_ERROR,
            error
          )
        );
      }
    });
  }

  /**
   * Set the precision (decimal places) for calculations
   * @param scale Number of decimal places (0-100)
   */
  async setPrecision(scale: number): Promise<void> {
    // Validate scale
    if (scale < 0 || scale > 100) {
      throw new BCCalculatorError(
        `Invalid precision: ${scale}. Must be between 0 and 100`,
        BCErrorCode.VALIDATION_ERROR
      );
    }

    // Set scale using BC's scale variable
    await this.evaluate(`scale=${scale}`);
    this.currentPrecision = scale;
  }

  /**
   * Get current precision setting
   */
  getPrecision(): number {
    return this.currentPrecision;
  }

  /**
   * Resolve a pending calculation request
   */
  private resolvePendingRequest(result: string): void {
    if (!this.pendingRequest) return;
    
    const { resolve, timer } = this.pendingRequest;
    if (timer) clearTimeout(timer);
    
    this.pendingRequest = null;
    resolve(result);
  }

  /**
   * Reject a pending calculation request
   */
  private rejectPendingRequest(error: Error): void {
    if (!this.pendingRequest) return;
    
    const { reject, timer } = this.pendingRequest;
    if (timer) clearTimeout(timer);
    
    this.pendingRequest = null;
    reject(error);
  }

  /**
   * Check if the process is available for new calculations
   */
  isAvailable(): boolean {
    return this.isReady && this.pendingRequest === null && this.process !== null;
  }

  /**
   * Kill the BC process
   */
  kill(): void {
    if (this.process) {
      try {
        this.process.kill('SIGTERM');
      } catch (error) {
        // Ignore errors when killing
      }
      this.process = null;
      this.isReady = false;
    }
  }

  /**
   * Get process ID (for debugging)
   */
  getPid(): number | undefined {
    return this.process?.pid;
  }
}